### PR TITLE
add support for 32-bit PNGs, refactoring

### PR DESCRIPTION
**Pull request by rfomin**

### CODE REVIEW

File: src/v_fmt.c [hunk 1]
```diff
@@ -315,63 +315,289 @@ static void *DummyFlat(int lump, pu_tag tag)
...
 
-    struct spng_ihdr ihdr = {0};
-    int ret = spng_get_ihdr(ctx, &ihdr);
+    int fmt;
+    switch (ihdr.color_type)
+    {
+        case SPNG_COLOR_TYPE_INDEXED:
+            fmt = SPNG_FMT_PNG;
+            break;
+        case SPNG_COLOR_TYPE_GRAYSCALE:
```
Comment (fabiangreffrath) [inline src/v_fmt.c:399]: Is this fall through intended?

Comment (rfomin) [reply src/v_fmt.c:399]: Yes, I convert everything to `SPNG_FMT_RGB8` or `SPNG_FMT_RGBA8` for simplicity.

**Comment by MrAlaux** [comment]
The missing pixels are still there, I'm afraid:
woof0001

I'm getting off my computer now, so here's a WAD containing the Wolf3D palette and a colormap for it so you can continue to test by yourself, if you'd like: [WolfPal.zip](https://github.com/user-attachments/files/16736318/WolfPal.zip)

**Comment by rfomin** [comment]
> The missing pixels are still there, I'm afraid

Hmm, I can't reproduce it:
woof0006

**Comment by MrAlaux** [comment]
> Hmm, I can't reproduce it

Maybe try going on Map32, whose green floor contrasts with the dog's colors, and check the front-facing rotation frames (can also use Pause and the freeze cheat for convenience).

**Comment by fabiangreffrath** [comment]
> > The missing pixels are still there, I'm afraid
> 
> Hmm, I can't reproduce it: woof0006

Wow, what's going on with the status bar and the outdoor area?

File: src/i_video.c [hunk 2]
```diff
@@ -1029,27 +1029,29 @@ byte I_GetPaletteIndex(byte *palette, int r, int g, int b)
...
     byte best;
     int best_diff, diff;
-    int i;
+    int i, dr, dg, db;
 
     best = 0;
     best_diff = INT_MAX;
 
     for (i = 0; i < 256; ++i)
     {
```
Comment (fabiangreffrath) [inline src/i_video.c:1038]: Looks very good! Do you have a benchmark for this?

**Comment by rfomin** [comment]
@MrAlaux fixed here: https://github.com/fabiangreffrath/woof/pull/1867/commits/abfc2283d4502c8c8159cd3088bab4d30af6363b Thanks!

woof0001

> Wow, what's going on with the status bar and the outdoor area?

That's Wolf3D palette.

File: src/i_video.c (hunk 2 above)
Comment (rfomin) [reply src/i_video.c:1038]: About twice as fast. I was just timing the conversion of a large 32 bit [TITLEPIC.zip](https://github.com/user-attachments/files/16736624/TITLEPIC.zip)

Comment (fabiangreffrath) [reply src/i_video.c:1038]: Impressive, thank you!

**Review by fabiangreffrath — APPROVED** [review]
Fantastic work, thank you very much!

**Pull request merged**